Panic when Ingester is starting

**Describe the bug**
When one of my Ingesters is starting, it crashloops with `panic: runtime error: slice bounds out of range [2610010201:824]`

**To Reproduce**
Not sure how this started or how to reproduce

**Environment:**
 - Infrastructure: Kubernetes
 - Deployment tool: Kustomize
 - Tempo version: 1.2.1

**Additional Context**
Full panic output:
```
level=info ts=2022-01-03T12:49:23.234802886Z caller=main.go:189 msg="initialising OpenTracing tracer"
level=info ts=2022-01-03T12:49:23.256529595Z caller=main.go:108 msg="Starting Tempo" version="(version=, branch=HEAD, revision=0eaca8a01)"
level=info ts=2022-01-03T12:49:23.364022024Z caller=server.go:260 http=[::]:3100 grpc=[::]:9095 msg="server listening on addresses"
level=info ts=2022-01-03T12:49:23.364817735Z caller=memberlist_client.go:394 msg="Using memberlist cluster node name" name=tempo-ingester-norwayeast-3-0-d7d51877
level=info ts=2022-01-03T12:49:23.364933437Z caller=module_service.go:64 msg=initialising module=server
level=info ts=2022-01-03T12:49:23.365034438Z caller=module_service.go:64 msg=initialising module=overrides
level=info ts=2022-01-03T12:49:23.364857736Z caller=module_service.go:64 msg=initialising module=store
level=info ts=2022-01-03T12:49:23.365201941Z caller=module_service.go:64 msg=initialising module=memberlist-kv
level=info ts=2022-01-03T12:49:23.365511945Z caller=module_service.go:64 msg=initialising module=ingester
level=info ts=2022-01-03T12:49:23.365535045Z caller=ingester.go:330 msg="beginning wal replay"
level=info ts=2022-01-03T12:49:23.365595546Z caller=wal.go:101 msg="beginning replay" file=57716e16-32a0-4e83-9b78-e7b161669887:single-tenant:v2:snappy:v1 size=6422528
level=info ts=2022-01-03T12:49:23.396842291Z caller=memberlist_client.go:506 msg="joined memberlist cluster" reached_nodes=7
level=info ts=2022-01-03T12:49:23.434710229Z caller=wal.go:128 msg="replay complete" file=57716e16-32a0-4e83-9b78-e7b161669887:single-tenant:v2:snappy:v1 duration=69.121083ms
level=info ts=2022-01-03T12:49:23.434804131Z caller=rescan_blocks.go:34 msg="beginning replay" file=57716e16-32a0-4e83-9b78-e7b161669887:single-tenant:v2:none: size=12275712
panic: runtime error: slice bounds out of range [2610010201:824]

goroutine 403 [running]:
github.com/google/flatbuffers/go.(*Table).GetVOffsetT(...)
        /drone/src/vendor/github.com/google/flatbuffers/go/table.go:134
github.com/google/flatbuffers/go.(*Table).Offset(0x0, 0x0)
        /drone/src/vendor/github.com/google/flatbuffers/go/table.go:16 +0xdf
github.com/grafana/tempo/pkg/tempofb.(*KeyValues).Key(0xc00024d848)
        /drone/src/pkg/tempofb/KeyValues.go:37 +0x25
github.com/grafana/tempo/pkg/tempofb.(*SearchBlockHeaderMutable).AddEntry(0xc04125dba0, 0xc00024d8a8)
        /drone/src/pkg/tempofb/SearchBlockHeader_util.go:30 +0xa9
github.com/grafana/tempo/tempodb/search.newStreamingSearchBlockFromWALReplay.func1({0xc0456bea90, 0x262e840, 0xc0456c0000})
        /drone/src/tempodb/search/rescan_blocks.go:92 +0x7c
github.com/grafana/tempo/tempodb/wal.ReplayWALAndGetRecords(0xc0005fa878, {0x269aad0, 0x3926a50}, 0x3b, 0xc00024dab0)
        /drone/src/tempodb/wal/replay.go:52 +0x335
github.com/grafana/tempo/tempodb/search.newStreamingSearchBlockFromWALReplay({0xc0001319f8, 0xc041296a80}, {0xc041290676, 0x3b})
        /drone/src/tempodb/search/rescan_blocks.go:90 +0x211
github.com/grafana/tempo/tempodb/search.RescanBlocks({0xc00011e380, 0x2630660})
        /drone/src/tempodb/search/rescan_blocks.go:38 +0x4dd
github.com/grafana/tempo/modules/ingester.(*Ingester).replayWal(0xc000806300)
        /drone/src/modules/ingester/ingester.go:337 +0x1c6
github.com/grafana/tempo/modules/ingester.(*Ingester).starting(0xc000806300, {0x2677ce0, 0xc040bbcd40})
        /drone/src/modules/ingester/ingester.go:101 +0x2c
github.com/grafana/dskit/services.(*BasicService).main(0xc0003e8820)
        /drone/src/vendor/github.com/grafana/dskit/services/basic_service.go:157 +0x78
created by github.com/grafana/dskit/services.(*BasicService).StartAsync.func1
        /drone/src/vendor/github.com/grafana/dskit/services/basic_service.go:119 +0xbe
```

--- a/tempodb/encoding/v2/page.go
+++ b/tempodb/encoding/v2/page.go
@@ -91,7 +91,10 @@ func unmarshalPageFromReader(r io.Reader, header pageHeader, buffer []byte) (*pa
 		buffer = buffer[:dataLength]
 	}
 
-	_, err = r.Read(buffer)
+	actualLength, err := r.Read(buffer)
+	if actualLength != len(buffer) {
+		return nil, fmt.Errorf("unexpected incomplete page read: expected:%d read:%d", len(buffer), actualLength)
+	}
 	if err != nil {
 		return nil, err
 	}

--- a/tempodb/encoding/v2/page_test.go
+++ b/tempodb/encoding/v2/page_test.go
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/binary"
 	"errors"
+	"fmt"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -142,4 +143,19 @@ func TestCorruptHeader(t *testing.T) {
 	page, err = unmarshalPageFromReader(bytes.NewReader(buffBytes), constDataHeader, nil)
 	assert.Nil(t, page)
 	assert.EqualError(t, err, "unexpected non-zero len data header")
+}
+
+func TestIncompletePageDetected(t *testing.T) {
+	dataLen := 100
+	del := 10
+	buff := &bytes.Buffer{}
+
+	_, err := marshalPageToWriter(make([]byte, dataLen), buff, constDataHeader)
+	require.NoError(t, err)
+
+	// Delete some bytes from the end of the page
+	buff.Truncate(buff.Len() - del)
+
+	_, err = unmarshalPageFromReader(buff, constDataHeader, nil)
+	require.EqualError(t, err, fmt.Sprintf("unexpected incomplete page read: expected:%d read:%d", dataLen, dataLen-del))
 }

